containerd-shim-spin-config.md

```diff
@@ -0,0 +1,235 @@
+# SKIP 00X - Configuring Runtime Options in the `containerd-shim-spin`
+
+Summary: Configuring runtime options in the `containerd-shim-spin`
+
+Owner: Kate Goldenring <kate.goldenring@fermyon.com> 
+
+Impacted Projects:
+
+- [x] spin-operator
+- [ ] `spin kube` plugin
+- [x] runtime-class-manager
+- [x] containerd-shim-spin
+- [ ] Governance
+- [ ] Creates a new project
+
+Created: 30/05/2024
+
+## Overview
+
+The `containerd-shim-spin` SpinKube executor currently lackd the ability to configure runtime flags for Spin applications. This proposal aims to address this by enabling three types of configurations: Spin execution settings, shim execution settings, and shim execution environment variables. These configurations will be set via Custom Resource Definitions (CRDs) in Kubernetes, specifically extending the SpinExecutor and SpinApp CRDs. This approach will allow operators to easily configure both the application and the platform execution environments.
+
+## Background
+
+SpinKube executors expose various options and arguments in order to configure how an application is executed. Take, for example, the following command:
+
+```sh
+spin up --listen "127.0.0.1:3000" --disable-pooling --log-dir /var/logs --runtime-config-file /var/spin/config.toml
+```
+
+The previous command tells the Spin runtime to do the following:
+
+- Listen for requests to the Spin app on IP address and port `127.0.0.1:3000`
+- Disable Wasmtime's pooling instance allocator
+- Use `/var/logs` as the log directory for the stdout and stderr of components
+- Use `/var/spin/config.toml` as the runtime configuration file.
+
+Currently, there is no way for configure any of these runtime flags through the `containerd-shim-spin`.
+
+Operators may also wish to configure Wasm runtime environment variables, such as specifying that the traces of a Spin app be emitted to a specific exporter address (analogous to `OTEL_EXPORTER_OTLP_ENDPOINT=http://127.0.0.1:4318 spin up`).
+
+A Runwasi containerd shim does more than execute a Wasm runtime such as Spin or Wasmtime or WasmEdge. It also pulls and prepares Wasm applications. The Shim may also have extra settings it would like to expose to configure how it prepares a Spin application. For example, a user should be able to disable pre-compilation for an application. The environment, such as the `RUST_LOG` level, of the shim should be configurable, too.
+
+In summary, the following are types of configuration that should be supported:
+
+- Wasm runtime specific options (oftentimes in the form of CLI flags: `spin up --listen "127.0.0.1:3000"`)
```

Comment:
I would call this spin configuration options.  When I think of wasm runtime options I think more of the underlying runtime options like `--disable-pooling`

Reply:
`wasmtime serve --addr` has the equivalent functionality.

Reply:
I see a distinction with `--addr` as more spin/wasmtime host than the underlying runtime configuration that set up the wasm runtime properties: https://github.com/bytecodealliance/wasmtime/blob/e97b74dc1027a8b4b051f3040e2e2793314c6390/crates/cli-flags/src/lib.rs#L40

Reply:
I see. I updated it to "Spin runtime". I would anticipate needing to configure the wasmtime runtime as well in the future -- it is just by happenstance that we can detect the need to `--disable-pooling`. I think these flags should also be configured via `SPIN_` env vars.